src/core/core/rpc_engine.cpp

```diff
@@ -112,20 +110,33 @@ namespace dsn {
             }
             else
             {
-                dassert(reply->get_count() == 0, 
-                    "reply should not be referenced by anybody so far");
-                delete reply;
+                if (reply)
+                {
+                    dassert(reply->get_count() == 0,
+                        "reply should not be referenced by anybody so far");
+                    delete reply;
+                }
                 return false;
             }
         }
-
+                
         dbg_dassert(call != nullptr, "rpc response task cannot be empty");
         if (timeout_task != task::get_current_task())
         {
             timeout_task->cancel(false); // no need to wait
         }
         timeout_task->release_ref(); // added above in the same function
-        
+
+        // if early terminate rpc  w/ empty reply
+        if (nullptr == reply)
+        {
+            call->set_delay(delay_ms);
+            call->enqueue(ERR_TIMEOUT, reply);
```

Comment:
I've considered this, but it will only complicate the error handling code in our upper layer, e.g., in replication. more error code we have, more complicated the code will be in upper layer.